diff --git a/llvm/cmake/modules/AddLLVM.cmake b/llvm/cmake/modules/AddLLVM.cmake
index a267166eb6c2..9bdf4963fef1 100644
--- a/llvm/cmake/modules/AddLLVM.cmake
+++ b/llvm/cmake/modules/AddLLVM.cmake
@@ -1581,6 +1581,8 @@ if(NOT LLVM_TOOLCHAIN_TOOLS)
     llvm-profdata
     llvm-profgen
     llvm-symbolizer
+    llvm-dis
+    llvm-dwarfutil
     # symlink version of some of above tools that are enabled by
     # LLVM_INSTALL_BINUTILS_SYMLINKS.
     addr2line
